tests/integration-tests/tests/iam/test_iam.py

```diff
@@ -363,6 +369,247 @@ def test_iam_resource_prefix(
+                "iam:PutRolePolicy",
+                "iam:ListAttachedRolePolicies",
+                "iam:ListInstanceProfiles",
+                "iam:ListInstanceProfilesForRole",
+            ]
+        )
+        statement["Resource"] = iam_role
+    elif "instance-profile" + iam_path + "*" in statement["Resource"].data["Fn::Sub"]:
+        statement["Action"].extend(
+            [
+                "iam:CreateInstanceProfile",
+                "iam:AddRoleToInstanceProfile",
+                "iam:TagInstanceProfile",
+                "iam:UntagInstanceProfile",
+            ]
+        )
+        statement["Resource"] = iam_role + iam_instance_profile
+
+
+def _update_existing_policy_in_permission_boundary(
+    permission_boundary_statements, iam_role, iam_instance_profile, iam_path
+):
+    """Update or remove sections of Policy statements from Permission Boundary"""
+    keep_first_cloudformation_policy = True
+    for statement in list(permission_boundary_statements):
+        if _find_resource_action("ec2:", statement["Action"]):
+            if not statement.get("Condition"):
+                permission_boundary_statements.remove(statement)
+        elif _find_resource_action("cloudformation:", statement["Action"]):
+            statement["Action"].clear()
+            statement["Action"].append("cloudformation:*")
+            keep_first_cloudformation_policy = False
+        elif _find_resource_action("iam:", statement["Action"]):
+            _change_existing_iam_policies(statement, iam_role, iam_instance_profile, iam_path)
+        elif (
+            (_find_resource_action("cloudformation:", statement["Action"]) and not keep_first_cloudformation_policy)
+            or _find_resource_action("logs:", statement["Action"])
+            or _find_resource_action("dynamodb:", statement["Action"])
+            or _find_resource_action("lambda:", statement["Action"])
+            or _find_resource_action("route53:", statement["Action"])
+            or _find_resource_action("imagebuilder:", statement["Action"])
+            or _find_resource_action("codebuild:", statement["Action"])
+            or _find_resource_action("tag:", statement["Action"])
+            or _find_resource_action("SNS:", statement["Action"])
+            or _find_resource_action("ssm:", statement["Action"])
+            or _find_resource_action("ecr:", statement["Action"])
+            or _find_resource_action("ec2messages:", statement["Action"])
+            or _find_resource_action("ssmmessages:", statement["Action"])
+            or _find_resource_action("kms:", statement["Action"])
+        ):
+            permission_boundary_statements.remove(statement)
+
+
+def _find_resource_action(resource_type, action_list):
+    """Check if a specific resource type exists in the Policy statement's Actions."""
+    return any(resource_type in action for action in action_list)
+
+
+def _create_iam_user_role_config(cluster_config, iam_resource_prefix):
```

Comment:
#### Major
1. In general, we should always  use Jinja to make runtime changes to a cloudformation template instead of ad-hoc "manual changes".
2. We can consider the following alternative approach instead of making direct changes to the user role template: implement a fixture that creates a permissions boundary policy according to the given IAM prefix and applies it to the user role

#### Minor
The value returned by this function is a stack template, not a generic config. I suggest to rename this function "_create_iam_user_role_template"

Reply:
Creating a new [PR](https://github.com/aws/aws-parallelcluster/pull/4782) to address new design changes